shrink.sh

```diff
@@ -0,0 +1,256 @@
+    is_lvm "$1"
+    status=$?
+    if [[ $status -ne 0 ]]; then
+        echo "Error: '$3' is not an LVM device"
+        return "$status"
+    fi
+    expected_size_in_bytes=$(parse_tag "$2")
+    if [[ -z "$expected_size_in_bytes" ]]; then
+        echo "Error: Tag $SHRINK_TAG not found for device '$3' in '$2'"
+        return 1
+    fi
+    check_filesystem_size "$1" "$expected_size_in_bytes"
+    status=$?
+    if [[ $status -ne 0 ]]; then
+        return "$status"
+    fi
+    check_volume_size "$1" "$expected_size_in_bytes"
+    status=$?
+    if [[ $status -ne 0 ]]; then
+        return "$status"
+    fi
+    is_device_mounted "$1"
+    status=$?
+    if [[ $status -ne 0 ]]; then
+        return "$status"
+    fi
+    shrink_volume "$1" "$expected_size_in_bytes"
+    return $?
+}
+
+function update_return_status_code() {
+    if [[ $1 -ne 0 ]]; then
+        run_status=$1
+    fi
+}
+
+function display_help() {
+    echo "Program to shrink an ext4 file system hosted in a Logical Volume. It retrieves the new size from the value of the option named systemd.shrinkfs captured in the /etc/fstab entry of the device. 
+    
+    Usage: '$(basename "$0")' [-h] [-d=|--device=|--all]
+
+    Example:
+
+    where:
+        -h show this help text
+        -d|--device= name or UUID of the device that holds an ext4 file system in /etc/fstab to shrink. It maps to the first column in the /etc/fstab file
+        --all processes all devices in the '/etc/fstab' that contains the $SHRINK_TAG option"
+
+}
+function parse_flags() {
+    for i in "$@"
+        do
+        case $i in
+            --all)
+            all_devices=true
+            ;;
+            -d=*|--device=*)
+            if [[ -n $device_name ]]; then
+                echo "Only one device flag '-d|--device' is supported"
+                exit 1
+            fi
+            device_name="${i#*=}"
+            ;;
+            -h)
+            display_help
+            exit 0
+            ;;
+            *)
+            # unknown option
+            echo "Unknown flag $i"
+            exit 1
+            ;;
+        esac
+    done
+    if [[ $all_devices == true && -n "$device_name" ]]; then
+        echo "Invalid combination of flags: --all and -d|--device"
+        exit 1
+    fi
+    if [[ $all_devices == false && -z "$device_name" ]]; then
+        display_help
+        exit 0
+    fi
+}
+
+function main() {
+
+    local device_name
+    local all_devices=false
+    local run_status=0
+    
+    parse_flags "$@"
+    IFS="${ARRAY_IFS}" devices=($(read_fstab))
+    if [[ ${#devices[@]} == 0 ]]; then
+        if [[ $all_devices == true ]]; then 
+            echo "No devices found in '/etc/fstab'"
+            exit 1
+        fi
+        echo "Device '$device_name' not found in fstab"
+        exit 1
+    fi
+  
+    for entry in "${devices[@]}"
+    do
+        device_name=$( get_device_name "$entry" )
+#        echo "entry is $entry" >&2
+        update_return_status_code $?
```

Comment:
Should we continue processing the current device if the step failed?

Reply:
but if we go for the exit on failure (`set -x`), that will reduce the line count quite a number :smile: